src/driver-gskill.c

```diff
@@ -0,0 +1,1365 @@
+ * it.
+ */
+static int
+gskill_reload_profile_data(struct ratbag_device *device)
+{
+	uint8_t buf[GSKILL_REPORT_SIZE_CMD] = { 0x0c, 0xc4, 0x0 };
+	int rc;
+
+	log_debug(device->ratbag, "Asking mouse to reload profile data\n");
+
+	rc = gskill_general_cmd(device, buf);
+	if (rc < 0) {
+		log_error(device->ratbag,
+			  "Failed to get mouse to reload profile data: %d\n",
+			  rc);
+		return rc;
+	}
+
+	return 0;
+}
+
+static int
+gskill_do_write_profile(struct ratbag_device *device,
+			struct gskill_profile_report *report)
+{
+	uint8_t *buf = (uint8_t*)report;
+	int rc;
+
+	report->checksum = gskill_calculate_checksum(buf, sizeof(*report));
+
+	rc = gskill_select_profile(device, report->profile_num, true);
+	if (rc < 0)
+		return rc;
+
+	/* Wait for the device to be ready */
+	msleep(200);
+
+	rc = ratbag_hidraw_raw_request(device, GSKILL_GET_SET_PROFILE,
+				       buf, sizeof(*report), HID_FEATURE_REPORT,
+				       HID_REQ_SET_REPORT);
+	if (rc < 0) {
+		log_error(device->ratbag,
+			  "Error while writing profile: %d\n", rc);
+		return rc;
+	}
+
+	rc = gskill_reload_profile_data(device);
+	if (rc)
+		return rc;
+
+	return 0;
+}
+
+static inline void
+gskill_read_resolutions(struct ratbag_profile *profile,
+			struct gskill_profile_report *raw)
+{
+	struct gskill_data *drv_data = ratbag_get_drv_data(profile->device);
+	struct gskill_profile_data *pdata =
+		&drv_data->profile_data[profile->index];
+	struct ratbag_resolution *resolution;
+	int dpi_x, dpi_y, hz, i;
+
+	log_debug(profile->device->ratbag,
+		  "Profile %d: DPI count is %d\n",
+		  profile->index, raw->dpi_num);
+
+	hz = GSKILL_MAX_POLLING_RATE / (raw->polling_rate + 1);
+
+	for (i = 0; i < raw->dpi_num; i++) {
+		dpi_x = raw->dpi_levels[i].x * GSKILL_DPI_UNIT;
+		dpi_y = raw->dpi_levels[i].y * GSKILL_DPI_UNIT;
+
+		resolution = ratbag_resolution_init(profile, i, dpi_x, dpi_y,
+						    hz);
+		resolution->is_active = (i == raw->current_dpi_level);
+		pdata->res_idx_to_dev_idx[i] = i;
+
+		ratbag_resolution_set_cap(resolution,
+					  RATBAG_RESOLUTION_CAP_SEPARATE_XY_RESOLUTION);
+	}
+}
+
+static int
+gskill_get_firmware_version(struct ratbag_device *device) {
+	uint8_t buf[GSKILL_REPORT_SIZE_CMD] = { 0x0c, 0xc4, 0x08 };
+	int rc;
+
+	rc = gskill_general_cmd(device, buf);
+	if (rc < 0) {
+		log_error(device->ratbag,
+			  "Failed to read the firmware version of the mouse: %d\n",
+			  rc);
+		return rc;
+	}
+
+	return buf[4];
+}
+
+static unsigned int
+gskill_misc_macro_code_to_keycode(uint8_t code)
+{
+	unsigned int keycode;
+
+	switch (code & 0x0f) {
+	case 0x0: keycode = KEY_LEFTCTRL;   break;
+	case 0x1: keycode = KEY_LEFTSHIFT;  break;
+	case 0x2: keycode = KEY_LEFTALT;    break;
+	case 0x3: keycode = KEY_LEFTMETA;   break;
+	case 0x4: keycode = KEY_RIGHTCTRL;  break;
+	case 0x5: keycode = KEY_RIGHTSHIFT; break;
+	case 0x6: keycode = KEY_RIGHTALT;   break;
+	case 0x7: keycode = KEY_RIGHTMETA;  break;
```

Comment:
This range looks *very* similar to the actual HID codes (redefined in libratbag-hidraw.c). I wonder if you couldn't reuse ratbag_hidraw_get_keycode_from_keyboard_usage() instead of having your own switch/case (doesn't seem to be working for BTN_*, sigh)